### ISSUE TITLE
Refactor CLI Results for get_htlc

### ISSUE THREAD

**Issue by ryanRfox** [bitshares/bitshares-core#1645]
**User Story**
As a `consumer of data from the CLI wallet's get_htlc response` I want `returned data to be more explicit` so that `I may more easily interrupt the information and act accordingly`.

**Additional Context**
The initial implementation of the CLI response for `get_htlc` is lacking in the ability for a user or system to efficiently consume and act on the returned data. 

First, the `expiration` field returns a sting indicating a time in the future. A computer system will likely prefer the UTC timestamp to act upon. A human likely prefers the existing string. Suggest to include both to expiration formats to accommodate both user types: `expiration_timestamp` and `expiration_string`.  

Second, the returned data could be refactored to reinforce the HTLC concept of a "conditional transfer" by organizing the JSON object into a hierarchy of:
  - id
  - transfer: to, from, amount, asset
  - conditions:
    - hash_lock: preimage_hash, preimage_length, hash_algo
    - time_lock: expiration_timestamp, expiration_string

Example:
```
{
  "database_id": "1.16.11",
  "transfer": {
    "from": "1.2.23833",
    "to": "1.2.792",
    "amount": 1000000,
    "asset": "1.3.0"
  },
  "conditions": {
    "hash_lock": {
      "preimage_hash": "8ca187c92a6a3892735ca9fdcc5af91f4f423ee8cda550158192cfe4219246ad",
      "hash_algo": "SHA256",
      "preimage_length": 19
    },
    "time_lock": {
      "expiration_timestamp": "2019-03-12T17:16:51"
      "expiration_string": "4 minutes in the future"
    }
  }
}
```

## CORE TEAM TASK LIST
- [ ] Evaluate / Prioritize Feature Request
- [ ] Refine User Stories / Requirements
- [ ] Define Test Cases
- [ ] Design / Develop Solution
- [ ] Perform QA/Testing
- [ ] Update Documentation

**Comment by jmjatlanta** [comment]
My initial take: We should make the names of any fields that are in both match for consistency. But I like the separation of things like "conditions" and "transfer". It breaks up the data into logical sections. Of course, I am open to other's opinions.

**Comment by jmjatlanta** [comment]
My suggestion for get_htlc:
```
{
  "id": "1.16.12",
  "transfer": {
    "from": "1.2.23833",
    "to": "1.2.792",
    "amount": {
      "amount": 10000000,
      "asset_id": "1.3.0"
     }
  },
  "conditions": {
    "hash_lock": {
      "preimage_hash": "57c3b1230e531a14e89b05cca71682081c9ce021aebe275b68cdb4fb8ffdf8f7",
      "hash_algo": "SHA256",
      "preimage_length": 13
    },
    "time_lock": {
      "expiration": "2019-03-15T15:50:27"
    }
  }
}
```

Summary of changes: 
Changed database_id to id
changed "amount" to a struct that matches get_object

**Comment by abitmore** [comment]
My suggestion for better human-readability:
get_htlc 1.16.12
{
  "id": "1.16.12",
  "transfer": {
    "from": "xeroc",
    "to": "john",
    "amount": "12.34567",
    "asset": "OPEN.BTC"
  },
  "conditions": {
    "hash_lock": {
      "preimage_hash": "57c3b1230e531a14e89b05cca71682081c9ce021aebe275b68cdb4fb8ffdf8f7",
      "hash_algo": "SHA256",
      "preimage_length": 13
    },
    "time_lock": {
      "expiration": "2019-03-15T15:50:27"
    }
  }
}
```

**Comment by abitmore** [comment]
If the requirement/specification is this, why not directly change the structure of `htlc_object` to something like below?
```
struct htlc_object {
  struct {
      account_id_type from;
      account_id_type to;
      share_type amount;
      asset_id_type asset_id;
   } transfer;
   struct {
      struct {
        preimage_hash;
        hash_algo;
        preimage_length;
      } hash_lock;
      struct {
         expiration;
      } time_lock;
   } conditions;
};
```
Then we don't need to convert. This is a new feature, there is no existing application or user, which means we can try to make it perfect better now than later.

**Comment by jmjatlanta** [comment]
> If the requirement/specification is this, why not directly change the structure of `htlc_object` to something like below?
...
> Then we don't need to convert. This is a new feature, there is no existing application or user, which means we can try to make it perfect better now than later.

Good suggestion. I am torn. The existing object is fairly flat, the fanciest thing being an embedded "htlc_hash" struct that encapsulates the hash with the hash algorithm that created it. Refactoring the structure will make the internal representation more closely reflect the human-readable output. But I do not know if we can completely remove the last-step reformatting due to some of the default serialization of some of the fields. i.e. expiration, the hash algo being stored as an enum but displayed as text, etc.

My pros/cons list: 
 - Keeping the struct as-is and reformatting it within the wallet
   - User Readability high
   - Code cleanliness medium
   - Machine readability medium
 - Modifying the struct, and doing (hopefully) less reformatting in the wallet
   - User Readability is the same
   - Code cleanliness medium 
     - less formatting in the wallet (possibly)
     - more hierarchy when filling / reading the struct (not sure if that is a good thing or bad thing)
   - Machine readability may be a bit worse, (i.e. extra dot notation)
 - Modifying the struct to do zero reformatting in the wallet
    - User Readability is poor (unless the json formatter is smarter/more controllable than I think it is)
    - Code cleanliness poor (i.e. loss of type checking)
    - Machine readability poor

These are just my current thoughts. I am not strongly opinionated between 1 or 2. 3 is dumb IMO.

**Comment by pmconrad** [comment]
I guess cli_wallet is being used by more tech-affine users and developers, who will be confused if cli output and wallet API differ too much. IMO restructuring the database object is an excellent idea.

**Comment by jmjatlanta** [comment]
I followed more or less the suggestion of @abitmore and received the following results:
```
get_htlc 1.16.15
{
  "id": "1.16.15",
  "transfer": {
    "from": "1.2.23833",
    "to": "1.2.792",
    "amount": 10000000,
    "asset_id": "1.3.0"
  },
  "conditions": {
    "hash_lock": {
      "preimage_hash": [
        2,
        "57c3b1230e531a14e89b05cca71682081c9ce021aebe275b68cdb4fb8ffdf8f7"
      ],
      "preimage_size": 13
    },
    "time_lock": {
      "expiration": "2019-03-16T22:07:30"
    }
  }
}
```
I believe this to be a lot closer to the desired result, with zero reformatting within the wallet. The deficiencies are:
- Account IDs instead of account names
- Asset ID instead of asset name
- preimage hash has cryptic hash algorithm
- preimage hash has no field labels
- perhaps preimage size should be part of the hash struct?

I will attempt to determine the best way to remedy these items.

**Comment by pmconrad** [comment]
`preimage_hash` being an array is due to the (arguably stupid) way `static_variant`s are represented in JSON. A generic solution would be nice, but breaks existing APIs and is out of scope here.
If `preimage_hash` has structure then `preimage_size` should not be part of it IMO, because the size of the preimage is not part of the hash but a separate piece of information.

**Comment by jmjatlanta** [comment]
Fixed by #1656
